Controllers and validation

MoJloDou-Proger opened the pull request:
Added:
- Classes UserController, FilmController (requests GET, POST, PUT)
- Classes User, Film
- Tests UserControllerTest, FilmControllerTest
- Validation for requests

MoJloDou-Proger committed:
Controllers and validation

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,68 @@
+
+    private void increaseFilmId(){
+        filmId++;
+    }
```
это вспомогательный метод, поэтому его хорошо было бы перенести в конец класса

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @PostMapping
+    public Film createFilm(@Valid @RequestBody Film film) throws ValidationException{
```
можно не указывать, что метод пробрасывает ValidationException т.к. этот эксепшен наследуется от RuntimeException, его необязательно обрабатывать

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,68 @@
+        validate(film);
+        films.add(film);
+        log.info("Фильм {} c id={} добавлен, объект: {}", film.getName(), filmId, film);
+        increaseFilmId();
```
хороший подход с отдельным методом, который отвечает за инкрементацию айдишника

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,68 @@
+                            "быть положительной");
+                } else throw new ValidationException("Дата релиза должна быть больше 28.12.1895");
+            } else throw new ValidationException("Описание не должно превышать 200 символов");
+        } else throw new ValidationException("Название не может быть пустым или null");
```
в целом, неплохая логика валидации, но вложенные if'ы сильно ухудшают восприятие кода, попробуй переделать, так, что бы как только не пройдет условие, будет выброшен эксепшен, и так, каждое условие друг за другом, а не вложенное одно в другое

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,68 @@
+                if (film.getReleaseDate().isAfter(MIN_DATE)){
+                    if (film.getDuration() <= 0) throw new ValidationException("Продолжительность фильма должна " +
+                            "быть положительной");
+                } else throw new ValidationException("Дата релиза должна быть больше 28.12.1895");
```
здорово, что для каждой проверки выводишь свое сообщение

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,68 @@
+        throw new ValidationException("Фильм с id=" + filmId + " не найден");
+    }
+
+    public void validate(Film film) throws ValidationException {
```
молодец, что вынес логику валидации в отдельный метод, 
но можно было бы пойти чуть дальше, и вынести ее в отдельный класс, 
тем самым будет больше разделение ответсвенности и такую логику будет чуть проще тестировать, т.к. можно будет просто протестировать отдельный класс

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,66 @@
+                } else throw new ValidationException("Дата рождения не может быть в будущем");
+            } else throw new ValidationException("Логин не может быть пустым и содержать пробелы");
+        } else throw new ValidationException("Почта не может быть пустой и должна содержать символ @");
+    }
```
здесь применимы все комментарии, относящиеся к валидационному методу в фильм контроллере

Fleaman777 commented on src/test/java/ru/yandex/practicum/filmorate/FilmControllerTest.java:
```diff
@@ -0,0 +1,80 @@
+public class FilmControllerTest {
+    private final LocalDate TEST_DATE = LocalDate.of(1895,12,29);
+    private final LocalDate MIN_DATE = LocalDate.of(1895,12,28);
+    FilmController filmController;
```
в тестах тоже хорошо придерживаться инкапсуляции

Fleaman777 commented on src/test/java/ru/yandex/practicum/filmorate/FilmControllerTest.java:
```diff
@@ -0,0 +1,80 @@
+        assertEquals("Фильм с id=9999 не найден", exception.getMessage());
+    }
+
+    public void checkException(Film film, String message){
```
супер, грамотное вынесение общей проверяющей логики в отдельный метод

Fleaman777 commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+	void contextLoads() {
+	}
+
+}
```
на этот класс можно не создавать тест, т.к. обычно в ApplicationClass нечего проверять)

Fleaman777 commented on src/test/java/ru/yandex/practicum/filmorate/UserControllerTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Test
+    public void dateOfBirthdayShouldNotBeInTheFuture() {
```
хорошие названия тест методов, содержат основную необходиимую информацию

MoJloDou-Proger committed:
- Валидация вынесена в класс validation.Validation, убрана вложенность if-ов
- Константы вынесены в класс constants.Constants

MoJloDou-Proger committed:
- Константы вынесены в constants.Constants
- Валидация вынесена в validation.Validation; убрана вложенность if-ов
- Тесты валидации собраны в классе ValidationTest

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/constants/Constants.java:
```diff
@@ -0,0 +1,8 @@
+public class Constants {
+    public static final LocalDate MIN_ALLOWED_DATE = LocalDate.of(1895,12,28);
+    public static final LocalDate DATE_BEFORE_MIN_ALLOWED = LocalDate.of(1895,12,27);
+}
```
отличное решение, с вынесением констант в отдельный класс

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/validation/Validation.java:
```diff
@@ -0,0 +1,35 @@
+        if (film.getDuration() <= 0) throw new ValidationException("Продолжительность фильма должна быть " +
+                "положительной");
+    }
+}
```
во, здорово, 
то, что я и хотел увидеть )

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/validation/Validation.java:
```diff
@@ -0,0 +1,35 @@
+import static ru.yandex.practicum.filmorate.constants.Constants.MIN_ALLOWED_DATE;
+
+public class Validation {
+    public static void userValidate(User user) {
```
поскольку метод выполняет действие, то его название должно начинаться с глагола, 
поэтому более правильное имя было бы: validateUser
аналагично с film
